Check and fix algorithms improvement

Djedouas opened the pull request:
## Description

Following https://github.com/qgis/QGIS/pull/59637 to improve the processing checks and fixes already here.

- avoid using new keyword to avoid memory leaks
- avoid using unnecessary pointers
- remove auto keyword
- better formatting
- better algorithm descriptions

Djedouas committed:
Improve angle check processing algorithm

- avoid using new keyword to avoid memory leaks
- remove auto keyword
- better formatting
- better algorithm description

Djedouas committed:
Improve area check processing algorithm

- avoid using new keyword to avoid memory leaks
- remove auto keyword
- better formatting
- better algorithm description

Djedouas committed:
Improve hole check processing algorithm

- avoid using new keyword to avoid memory leaks
- remove auto keyword
- better formatting
- better algorithm description

Djedouas committed:
Improve missing vertex check processing algorithm

- avoid using new keyword to avoid memory leaks
- remove auto keyword
- better formatting
- better algorithm description

Djedouas committed:
Improve angle fix processing algorithm

- avoid using unnecessary pointers
- remove auto keyword
- better formatting

Djedouas committed:
Improve area fix processing algorithm

- avoid using unnecessary pointers
- remove auto keyword
- better formatting

Djedouas committed:
Improve hole fix processing algorithm

- avoid using unnecessary pointers
- remove auto keyword
- better formatting

Djedouas committed:
Improve missing vertex fix processing algorithm

- avoid using unnecessary pointers
- remove auto keyword
- better formatting

DelazJ commented on src/analysis/processing/qgsalgorithmcheckgeometryangle.cpp:
```diff
@@ -76,27 +77,36 @@ void QgsGeometryCheckAngleAlgorithm::initAlgorithm( const QVariantMap &configura
+  ) );
+
+  std::unique_ptr<QgsProcessingParameterNumber> tolerance = std::make_unique<QgsProcessingParameterNumber>(
+    QStringLiteral( "TOLERANCE" ), QObject::tr( "Tolerance" ), Qgis::ProcessingNumberParameterType::Integer, 8, false, 1, 13
```
As mentioned in https://github.com/qgis/QGIS-Documentation/pull/9773, afaiu I find this tolerance parameter (_8 meaning 10^-8_) non-obvious, and its min/max values counter-intuitive. A parameter help is at least required for end users, but imho, we should move to something simpler as a plain double value with unit selector, as commonly used for this kind of parameter.

lbartoletti replied:
Yes but it's to keep consistency with Geometry Checker plugin:


https://github.com/user-attachments/assets/19d7327c-0c7c-4d07-bd29-35a51ebd8f1c

lbartoletti replied:
btw, Geometry Checker accepts only an integer https://github.com/qgis/QGIS/blob/master/src/analysis/vector/geometry_checker/qgsgeometrycheckcontext.cpp#L20

DelazJ replied:
> Yes but it's to keep consistency with Geometry Checker plugin:

Probably, but 1E-8 as shown in the plugin looks much more meaningful than 8. I don't know how strong is/should be the link between the two tools and I'm not speaking in terms of code, rather as a Processing end user who would expect some obviousness in the options and harmonization (and that we take advantage from improved parameter types in Processing)...

Djedouas replied:
Yes I agree that from an end user perspective, it could be confusing. However, we are tied with how the geometry checker is written right now.

We can't use the distance widget parameter because it would be nonsense and difficult to restrict the user to specific logarithmic values like "10, 1, 0.1, 0.001, 0.0001, etc.".

Our options are 
1. to create another parameter type, think of its UI, and that handles the differences we have between geographic and metric distances
2. to change how the geometry checker tolerance is given

I prefer to go with option 2. but in any case, it will be a work that would be done in a dedicated PR.

Right now, we could improve the documentation, or add a description of the tolerance parameter in every processing description.

DelazJ replied:
> Right now, we could improve the documentation, or add a description of the tolerance parameter in every processing description.

I would have said 
> Right now, we could improve the documentation, ~~or~~**AND** add a description of the tolerance parameter in every processing description.

They should have a start of answer in the GUI imho.

Djedouas replied:
Added a short text in every processing short help.
> The "Tolerance" advanced parameter defines the numerical precision of geometric operations, given as an integer n, meaning that any difference smaller than 10⁻ⁿ (in map units) is considered zero.

DelazJ replied:
Given that this option is in advanced panel, I doubt people will modify it and don't think it deserves a mention in the algorithm main help. It should instead be as a tooltip over the parameter only, i.e.:
```
  tolerance->setHelp( QObject::tr( "The \"Tolerance\" defines the numerical precision of geometric operations, "
                                   "given as an integer n, meaning that any difference smaller than 10⁻ⁿ (in map units) is considered zero." );
```

DelazJ reviewed (commented):
Hi @Djedouas 
Thanks for these updates. I was about to open issue reports for these and saw your PR, hence I'm asking here.

Djedouas committed:
Make angle check processing output layer optional

Djedouas committed:
Make area check processing output layer optional

Djedouas committed:
Make contained check processing output layer optional

Djedouas committed:
Make dangle check processing output layer optional

Djedouas committed:
Make degenerate polygon check processing output layer optional

Djedouas committed:
Make duplicate check processing output layer optional

Djedouas committed:
Make duplicate nodes check processing output layer optional

Djedouas committed:
Make gap check processing output layer optional

Djedouas committed:
Make follow boundaries check processing output layer optional

Djedouas committed:
Make line intersection check processing output layer optional

Djedouas committed:
Make hole check processing output layer optional

Djedouas committed:
Make missing vertex check processing output layer optional

Djedouas committed:
Make line layer intersection check processing output layer optional

Djedouas committed:
Make overlap check processing output layer optional

Djedouas committed:
Make multipart check processing output layer optional

Djedouas committed:
Make point covered by line check processing output layer optional

Djedouas committed:
Make segment length check processing output layer optional

Djedouas committed:
Make point in polygon check processing output layer optional

Djedouas committed:
Make self intersection check processing output layer optional

Djedouas committed:
Make self contact check processing output layer optional

Djedouas committed:
Make sliver polygon check processing output layer optional

Djedouas committed:
fix(algorithmcheckgeometry): add const for geometry error loop

Djedouas committed:
fix(algorithmfixgeometry): add verification of feedback canceled

Djedouas commented on src/analysis/processing/qgsalgorithmcheckgeometryangle.cpp:
```diff
@@ -76,27 +77,36 @@ void QgsGeometryCheckAngleAlgorithm::initAlgorithm( const QVariantMap &configura
+  ) );
+  addParameter( new QgsProcessingParameterFeatureSink(
+    QStringLiteral( "OUTPUT" ), QObject::tr( "Output layer" ), Qgis::ProcessingSourceType::VectorAnyGeometry
+  ) );
```
That's right, it has been resolved in the last commits.

DelazJ commented:
> What do you think of changing the "Check geometry (\<name>)" processing name to simply \<name>" ?

I agree that it looks crowded. I'm not very sure that an alg named "Angle" or "Contained" would be obvious (well, it is still displayed under the check group).
Maybe "Check \<name\>"? though I'm still not sure if "Check contained" is a quite expressive name...

PS: I'm not sure that the alg name is supposed to be in title case... ("Check Geometry (Dangle) --> Check geometry (dangle)")
PS2: most of the time, the alg outputs are named using the alg name, i.e., Buffer --> Buffered, while here they all output "layer output" and "Error layer". It could be nice to have part of the alg name in the output(s)
PS3: Thanks for the switch to optional output.

Djedouas committed:
fix(algorithmcheckgeometry): avoid memory leaks

QList<QgsGeometryCheckError *> contains pointers to non-Qt objects.
When the list is deleted, all pointers inside the list are deleted, but
every data that each pointer points to is not automatically freed.

Djedouas commented:
>  Maybe "Check <name>"? though I'm still not sure if "Check contained" is a quite expressive name...
> PS: I'm not sure that the alg name is supposed to be in title case... ("Check Geometry (Dangle) --> Check geometry (dangle)")

@DelazJ In fact, I think that we need to keep the long algorithm name, because when used inside a processing model we must see a full name. I really don't have an idea about title case or not for the name, that's why there is a mix... What do you think about this?

> PS2: most of the time, the alg outputs are named using the alg name, i.e., Buffer --> Buffered, while here they all output "layer output" and "Error layer". It could be nice to have part of the alg name in the output(s)

That's a good point. We could add the check name like the algorithm name:
- Errors layer (check dangle)
- Output layer (check dangle)

DelazJ commented:
> In fact, I think that we need to keep the long algorithm name, because when used inside a processing model we must see a full name. 

@Djedouas Sorry I do not get the point with the modeler.
IMHO No need to repeat the group name in the alg, which is already displayed at the top of the alg dialog. 
![image](https://github.com/user-attachments/assets/5fa5b69b-83b3-4824-8d32-9b1d0f234bc0)

The name should really be shortened for the reason you mentioned earlier, and maybe revisited to greatly reflect what the alg does. Do we really need the "check..." every time? What about e.g.:
* Check geometry (Angle) --> Find(/Detect) small(er?) angles
* Fix geometry (Angle) --> Delete small(er?) angles

  > I really don't have an idea about title case or not for the name, that's why there is a mix... What do you think about this?

Actually, they should not be in title case. If you walk through the Processing toolbox you'll notice that only the first letter is in upper case, unless there is an acronym. This is the rule to follow.
> That's a good point. We could add the check name like the algorithm name:

>    Errors layer (check dangle)
>    Output layer (check dangle)

I'm not really fan; still long and not inline with existing practice. What about e.g.:
* Dangle errors / Errors (dangle)
* Dangle / Dangle locations -- the global target of this output  is not always really clear to me
* Dangle fixed / Undangled (assuming it is English)

Sorry to add: There is also that great function to display algs short help when you hover over them in the toolbox. That may be helpful. But I realize there are quite a few algorithms using it. 😢 

If usual Processing algs reviewers could chime in...

Djedouas committed:
fix(algorithmgeometry): add tolerance help tooltip

Djedouas commented:
@DelazJ I will start renaming the algorithms with clearer names, like "Detect small angles", and output layer, like "Small angle errors".

> Sorry to add: There is also that great function to display algs short help when you hover over them in the toolbox. That may be helpful. But I realize there are quite a few algorithms using it. 😢

Do you mean implementing the `QgsProcessingAlgorithm::shortDescription()` method?

Djedouas committed:
fix(qgsalgorithmcheckgeometrypointcoveredbyline): remove duplicate output

Djedouas committed:
fix(qgsalgorithmcheckgeometryangle): remove duplicate output

Djedouas committed:
fix(qgsalgorithmcheckgeometrypointinpolygon): remove duplicate output

Djedouas committed:
chore(qgsalgorithmcheckgeometrymultimart): add test and compatibility with multipoint layers

Djedouas committed:
chore(check-fix-algorithms): improve names and help strings

Djedouas commented:
@DelazJ I tried not to add "Detect" to any new processing name, as it would be added to every name in the Check geometry section...

I prefer renaming the older from "Line intersections" to "Intersection (lines)" because there already is an "Intersection (multiple)" algorithm

Before

![image](https://github.com/user-attachments/assets/21762af5-39a9-45cb-9727-502a7acc7351)

After

![image](https://github.com/user-attachments/assets/baee6812-c0de-48e0-8569-570a0af1d734)

lbartoletti commented:
Is it ok @DelazJ ? Would like to merge before the freeze (and finish this QEP :) ).

DelazJ commented:
> Is it ok @DelazJ ?

@lbartoletti I don't have resources to review this PR so as said above I'm fine with what I've seen so far. I think tweaks with wording will come when using or from documentation repo.

With regard to renaming the algorithm, I'm not very keen. I find it confusing/misleading to try to mimic the (multiple) formatting:
- "Intersection" and "Intersection (multiple)" are two close algs that take common portions of features across layers, and they accept point, polygon and **line** layers. And they are coherent with the couples "Union - Union (multiple)" and "Difference - difference (multiple)".
- the existing "Line intersections" alg does something quite different, creating point where features cross. With the renaming (_and placement_), people can think it does the same as above but only for lines. What about "Point at lines intersection" (not sure where the plural is supposed to be)?

lbartoletti commented:
> @lbartoletti I don't have resources to review this PR

@DelazJ don't worry, it's not about the review of this PR (I can guarantee the code is ok), but more for wording/user point of view :)

DelazJ commented:
Yes, thanks @Djedouas

lbartoletti commented:
Many thanks @DelazJ for the review!
